src/internal/components/ScaledTextInput.tsx

```diff
@@ -0,0 +1,53 @@
+import { useEffect, useRef, useState } from 'react';
+import { cn } from '../../styles/theme';
+import type { TextInputReact } from './TextInput';
+import { TextInput } from './TextInput';
+
+export function ScaledTextInput(props: TextInputReact) {
+  const inputContainer = useRef<HTMLDivElement | null>(null);
+  const offscreenRef = useRef<HTMLSpanElement | null>(null);
+  const [containerScale, setContainerScale] = useState(1);
+
+  useEffect(() => {
+    if (props.value && offscreenRef.current && inputContainer.current) {
+      const containerWidth = inputContainer.current.offsetWidth;
+      const textWidth = offscreenRef.current.offsetWidth;
+      const scaleFactor = containerWidth / textWidth;
+
+      if (scaleFactor < 1) {
+        setContainerScale(scaleFactor);
+      } else {
+        setContainerScale(1);
+      }
+    } else {
+      setContainerScale(1);
+    }
+  }, [props.value]);
```

Comment:
Open question, could this re-fire too many times if we keep `props.value` too wide?